### ISSUE TITLE
Duplicate records found in UNIQUE_HASH_INDEX

### ISSUE THREAD

**Issue by masimakopoulos** [orientechnologies/orientdb#3556]
Hello,

I have a Vertex called Contact that has a field Id that is using a UNIQUE_HASH_INDEX. 
```
CREATE PROPERTY Contact.Id STRING
CREATE INDEX Contact.Id UNIQUE_HASH_INDEX
```
Using the binary protocol I am writing/reading/deleting contacts to a database from 6 different machines running in parallel, created two days ago with a few hundred thousand records (empty db before creating indexes). After noticing some weird behavior where I used an indexed key to retrieve a record but was getting no results, I went ahead and clicked rebuild indexes. 8 out of 10 indexed (all unique hash) failed as they contained duplicate records, including contact id. I executed a group by id query 
```
select Id, count(*) as num, list(@rid) from Contact group by Id order by num desc
```
that revealed that I had 2,3 and 4 times the same id on contact. 

I am using Orient 2.0.1 in standalone mode.

**Comment by laa** [comment]
Hello @mxa055 ,
Do you use tx mode or  you work in non tx mode ?

**Comment by mxa055** [comment]
I am using transactions for all write operations

**Comment by mxa055** [comment]
Don't know if it helps at all, but after the index rebuild failed I noticed the following behavior:

When selecting contact by Id using where Id='one of the duplicates' I got 0 results back. 
When executing the group by mentioned above I could see 4 different records with that Id.
Inserting yet another record with that same Id from the web studio succeeded where id select was returning just this record, but group by was returning all 5 records.
Trying to insert a 6th record with the same id gave me a unique index exception.

**Comment by laa** [comment]
Is it possible to send me your db, I will try to reproduce it ?

**Comment by mxa055** [comment]
I started again from an empty database for debugging purposes. As soon as I get it to a similar state I will send it over.

**Comment by mxa055** [comment]
Where should I send it?

**Comment by laa** [comment]
What is your db size ?

On Wed, Feb 11, 2015 at 1:11 PM, Michael <notifications@github.com> wrote:

> Where should I send it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/orientechnologies/orientdb/issues/3556#issuecomment-73865338>
> .
>



-- 
Best regards,
Andrey Lomakin.

**Comment by mxa055** [comment]
20Mb give or take. I can email you a dropbox link if you 'd like.

**Comment by laa** [comment]
OK.
Lets do it.

On Wed, Feb 11, 2015 at 1:35 PM, Michael <notifications@github.com> wrote:

> 20Mb give or take. I can email you a dropbox link if you 'd like.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/orientechnologies/orientdb/issues/3556#issuecomment-73867956>
> .
>



-- 
Best regards,
Andrey Lomakin.

**Comment by mxa055** [comment]
Done. Let me know if everything is ok.

**Comment by mxa055** [comment]
I reproduced the error using UNIQUE sbtree index today as well, so it must be a more general error affecting both unique index algorithms.

**Comment by mxa055** [comment]
I have some more info on this. What I found out happens is that the indexes fail to rebuild causing 
```
select from Contact where Id='1234'  
```
to return no records. In contrast doing:
```
select from Contact where Id like `1234`
```
returns the record. When the index is broken after a failed rebuild I can insert the same record again (which is what will happen anyway since I could not retrieve it through it's id to update it thus going for an insert operation). 
Now doing a where id = '1234' will retrieve the last entered record, but where id like `1234` will retrieve two records. From then on when ever an index rebuild is triggered any record that is accessed to be updated will be duplicated.

I do not know what causes the original index corruption, but as is the database is unusable since it is totally unreliable for existing data and has 10s of duplicates of each record with inconsistent edges between nodes with the same data and id but different orids. 

@lvca is there anything we can do about this currently, as this version is in production already and real data is being corrupted as we speak.

**Comment by lvca** [comment]
@mxa055 this error is pretty critical. I saw you used `'` and `` in your examples. Is it a case or could it be the cause?

**Comment by mxa055** [comment]
Sorry for that, I was just duck typing here, not copy pasting from actual examples.

**Comment by mxa055** [comment]
Here are some screenshots from actual examples:

Duplicate records of unique index Contact.Id
[screenshot link] (http://www.dropbox.com/s/n4erwm7ihe0wtsf/2015-02-13%2012_25_15-Mozilla%20Firefox.png)

After adding a contact is added to a broken index with an existing id (like vs =)
[screenshot link] (http://www.dropbox.com/s/xjgdhfvbg2jqw69/2015-02-13%2012_26_33-Mozilla%20Firefox.png?dl=0)

Trying to rebuild the index:
[screenshot link] (http://www.dropbox.com/s/2z2gr9grxp9jhcz/2015-02-13%2012_27_26-.png?dl=0)

Like vs = on the broken index (I can now add a contact with the same id again)
[screenshot link] (http://www.dropbox.com/s/5edx3qhkfhxx7ol/2015-02-13%2012_27_54-Mozilla%20Firefox.png?dl=0)

**Comment by tglman** [comment]
hi @mxa055, can you resend the database also to me at 'tglman - at- orientechnologies.com'?
anyway, how do you reproduce this ? is it your case a multiclient load test? i would like to have determinist way to reproduce it, for avoid regression in future.

**Comment by mxa055** [comment]
hey @tglman I just emailed you a copy of the db and logs I sent to @laa. Our environment has a web api where changes to the orient happen from the user and 6-10 workers (other machines, same db user) that alter the data without user interaction. You will see in the logs that we get quite a few Concurrency exceptions that we resolve in our code due to the parallel editing of data (not sure if this has something to do with the index being corrupted in the first place).
Any more details you need feel free to ask or email me.

**Comment by tglman** [comment]
hi @mxa055 can you try to send it also to support@orientechnologies.com i didn't receive the mail till now.

**Comment by mxa055** [comment]
probably a typo on the previous email. Resend it to your email plus the support one. Let me know if you still didn't got them.

**Comment by tglman** [comment]
hi all,

I did some investigation i saw the duplicate in the sent database, and i could also reproduce the issue, but only in high concurrency with multiple clients, so it's not so easy detect the actuall point where the problem is originated, working on it .

**Comment by masimakopoulos** [comment]
I am actually seeing this even with just two clients that get a few concurrency exceptions as trying to modify the same records at some point.

**Comment by tglman** [comment]
@masimakopoulos 

now i'm really struggly to reproduce this issue, i did it once, but not anymore.
can you try to write a test case that reproduce the issue?
i did one and let it run for hours with no success

**Comment by lvca** [comment]
@masimakopoulos I'm interested to see this issue fixed, but reproducing it seems quite hard. Please could you help us on this? Thanks.

**Comment by masimakopoulos** [comment]
I will try to put something together @lvca. I am going to use the .Net driver and C# code if that is ok, since that is what I am more familiar with.

**Comment by masimakopoulos** [comment]
Ok I now have something. It happens much easier than I thought. I am using the latest version of the .Net binary driver and I am using transactions for everything.
DB startup: https://db.tt/BRtbnMKA v2.0.1
Use LightWeightEdges Enabled.

I create a class called Profile with a field Id that uses UNIQUE_HASH_INDEX. I add a profile with Id=1 in a transaction and commit the transaction. I then try to add another profile using the same Id and I get 

com.orientechnologies.orient.core.storage.ORecordDuplicatedException: Cannot index record #12:1: found duplicated key '1' in index 'Profile.Id' previously assigned to the record #12:0
BUT:
doing a select from Profile in the studio I get both records:
https://db.tt/hO5eaR9c
I get a corrupted index
https://db.tt/HW8kpWf3
And this is the output on the db logs when this happens
https://db.tt/In2terG3

So just trying to add a record with an existing Id inside a transaction creates a bad index state. When the index is tried to be rebuild, everything breaks.

Here is the code:
https://www.dropbox.com/s/yg5npsduhkbj5q1/Orient.zip?dl=0

**Comment by tglman** [comment]
hi @masimakopoulos 

Thank you this help us a lot, i home to get back with an answer quickly.

**Comment by tglman** [comment]
found the bug! and is not a bug, is more a setting
from the .NET client the transaction log was off
CommitTransaction.cs line 42
`request.AddDataItem((byte)0); // use log 0 = no, 1 = yes`

changing it to:
`request.AddDataItem((byte)1); // use log 0 = no, 1 = yes`

everything work properly

can i close it ?

**Comment by lvca** [comment]
@tglman great! I'm not sure if this remains a problem. Why the log affects this?

**Comment by masimakopoulos** [comment]
@tglman great that you found the bug. @lvca I think that no matter what the client does the db should be consistent. Maybe this behaviour ought to be rethinked.

**Comment by lvca** [comment]
Ok, moved forward to 2.2

**Comment by laa** [comment]
Actually it is super simple to fix by removing this feature it made and makes system fragile.

**Comment by laa** [comment]
I do not see reason how tx can be tx if it does not support durability and data consistency features.

**Comment by lvca** [comment]
The idea of log yes/no was to use tx to flush a block of updates via network.

**Comment by laa** [comment]
and that makes system fragile and unstable in case of server crash and violations of data consistency.

**Comment by masimakopoulos** [comment]
@lvca I understand your point of view, of this being a feature rather than a bug, where you just need something faster for batch operations inside transactions, but I have to agree with @laa that when using transactions data consistency is of the utmost importance and no matter what the client or user does, the db should not be left into an inconsistent state where data loss is more than probable. Turning on log as a quick fix for now seems might be acceptable, but I think a better solution should be provided. Also since this **is the default setting of the .Net driver** I bet there are a lot of deployments that have this "bug" right know, probably with them not even knowing them.

**Comment by lvca** [comment]
That setting must be true by default with all drivers. In some cases you need a massive insertion and you don't use indexes, then setting it to false helps. I don't want to remove it, but rather add a line in documentation that explain this better.
